File: app/src/main/java/com/codeprogression/mvpb/core/BaseBindingPresenter.java
package com.codeprogression.mvpb.core;

import android.databinding.Observable;
import android.support.annotation.CallSuper;

import rx.subscriptions.CompositeSubscription;

public abstract class BaseBindingPresenter<VM extends Observable> {

    private CompositeSubscription subscriptions;
    private VM viewModel;

    @CallSuper
    public final void attach(VM viewModel) {
        if (this.viewModel == null) {
            this.viewModel = viewModel;
        }
        resume();
    }

    @CallSuper
    public final void detach() {
        this.viewModel = null;
        pause();
    }

    protected final boolean hasViewModel(){
        return viewModel != null;
    }

    public final VM getViewModel() {
        return viewModel;
    }

    private void resume() {
        if (subscriptions != null) {
            subscriptions.unsubscribe();
        }

        this.subscriptions = new CompositeSubscription();
        load();
    }

    private void pause() {
        unload();
        if (subscriptions != null) {
            subscriptions.unsubscribe();
        }
    }


    protected abstract void load();

    protected void unload(){

    }

}

File: app/src/main/java/com/codeprogression/mvpb/core/BaseBindingPresenter.kt
package com.codeprogression.mvpb.core

import android.databinding.Observable
import android.support.annotation.CallSuper
import rx.Subscription
import rx.subscriptions.CompositeSubscription

abstract class BaseBindingPresenter<VM : Observable> {

    private var subscriptions: CompositeSubscription = CompositeSubscription()
    lateinit protected var viewModel: VM

    protected var hasViewModel: Boolean = false
        private set

    @CallSuper
    fun attach(viewModel: VM) {
        hasViewModel = true
        this.viewModel = viewModel
        resume()
    }

    @CallSuper
    fun detach() {
        hasViewModel = false
        pause()
    }

    private fun resume() {
        subscriptions.unsubscribe()
        subscriptions = CompositeSubscription()
        load()
    }

    private fun pause() {
        unload()
        subscriptions.unsubscribe()
    }

    protected abstract fun load()

    protected fun unload() {

    }

    fun addSubscription(subscription: Subscription) {
        subscriptions.add(subscription)
    }

    fun removeSubscription(subscription: Subscription) {
        subscriptions.remove(subscription)
    }

}

